Adding readme for s3 source [KCON-9] (#376)
Aiven-Open/cloud-storage-connectors-for-apache-kafka: s3-source-connector/README.md

# Aiven's S3 Source Connector for Apache Kafka

> This is a source Apache Kafka Connect connector that stores AWS S3 bucket objects in Apache Kafka.

Reviewer: ```suggestion
This is a source Apache Kafka Connect connector that adds AWS S3 bucket objects onto an Apache Kafka topic.
```

   {`aws.sts.role.arn`, `aws.sts.role.session.name`} blank. In case you prefer to build your own custom
   provider, pass the custom provider class as a parameter to `aws.credential.provider`

It is important not to use both 1 and 2 simultaneously.
Using option 2, it is recommended to specify the S3 bucket region in `aws.s3.region` and the
corresponding AWS STS endpoint in `aws.sts.config.endpoint`. It's better to specify both or none.
It is also important to specify `aws.sts.role.external.id` for the security reason.
(see some details [here](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-user_externalid.html)).

### File name format

> File name format is tightly related to [Record Grouping](#record-grouping)

The connector uses the following format for input files (blobs):
`<prefix><filename>`.

`<prefix>`is the optional prefix that can be used, for example, for
subdirectories in the bucket.
`<filename>` is the file name. The connector has a fixed
template for file names.

    Fixed template for file : `{{topic}}-{{partition}}-{{start_offset}}`

Example object name : customertopic-00001-1734445664111.txt

> ## Data Format

Reviewer: This header seems incorrect for what is below.
Is this just # Configuration ?

Author: Yea, it is just configuration, but under it there are relevant sections. 
And similar to readme in sink.

Connector class name, in this case: `io.aiven.kafka.connect.s3.AivenKafkaConnectS3SourceConnector`.

### S3 Object Names

S3 connector stores series of files in the specified bucket.

> Each object is named using pattern `[<aws.s3.prefix>]<topic>-<partition>-<start_offset>-<extension>`

Reviewer: I dont think we have it set yet, but there should be a configuration for setting object names.

Author: I have rephrased it. But we are still following the pattern [<aws.s3.prefix>]<topic>-<partition>-<start_offset>-<extension> right ?

This means that in case you have the record and key schema like:

Key schema:
```json
{
  "type": "string"
}
```

Record schema:
```json
{
    "type": "record", "fields": [
      {"name": "foo", "type": "string"},
      {"name": "bar", "type": "long"}
  ]
}
```
the final `Avro` schema for `Parquet` is:
```json
{
    "type": "record", "fields": [
      {"name": "key", "type": "string"},
      {"name": "offset", "type": "long"},
      {"name": "timestamp", "type": "long"},
      {"name": "headers", "type": "map", "values": "long"},
      { "name": "value",
        "type": "record",
        "fields": [
          {"name": "foo", "type": "string"},
          {"name": "bar", "type": "long"}
        ]
      }
  ]
}
```
**NB!**

- Connector works just fine with and without Schema Registry

## Usage

### Connector Configuration

> **Important Note** Since this connector is developed aligning it with S3 sink connector,
> and since version `2.6`, all existing configuration in S3 sink
is deprecated and will be replaced with new one during a certain transition period (within 2-3 releases). Most of the
> configuration parameters remain same.

> List of deprecated configuration parameters:

Reviewer: Lets remove the deprecated config params and not support them in the S3 Source Connector so we dont have any legacy we need to clean up in the future.

[here](https://kafka.apache.org/documentation/#connect_resuming), about
the connector Configuration.

Here is an example connector configuration with descriptions:

```properties
### Standard connector configuration

## Fill in your values in these:

## These must have exactly these values:

# The Java class for the connector
connector.class=io.aiven.kafka.connect.s3.AivenKafkaConnectS3SourceConnector

# Number of worker tasks to run concurrently
tasks.max=1

# The key converter for this connector
key.converter=org.apache.kafka.connect.storage.StringConverter

# The type of data format used to write data to the kafka events.
# The supported values are: `jsonl`, 'avro', `parquet` and 'bytes'.
input.type=jsonl

# A comma-separated list of topics to use as output for this connector
# Also a regular expression version `topics.regex` is supported.
# See https://kafka.apache.org/documentation/#connect_configuring
topics=topic1,topic2

# A comma-separated list of topic partitions where the connector's offset storage reader
# can read the stored offsets for those partitions. If not mentioned, s3 objects will be read again if
# available in the bucket

> topic.partitions=1,2,3

Reviewer: This seem incorrect for the source connector. We should just read all the objects (ideally just once) is this a sink only config?

Author: This is for source connector, and based on these partitions, the offset storage reader reads the map

### Connector-specific configuration
### Fill in you values
# AWS Access Key ID
aws.access.key.id=YOUR_AWS_KEY_ID

# AWS Access Secret Key
aws.secret.access.key=YOUR_AWS_SECRET_ACCESS_KEY

#AWS Region
aws.s3.region=us-east-1

#The name of the S3 bucket to use
#Required.
aws.s3.bucket.name=my-bucket

#The prefix of the S3 bucket to use
#Optional.
aws.s3.prefix=file-prefix

#Errors tolerance
# Possible values 'none' or 'all'. Default being 'none'
# If 'none', then any errors during data transformations or reading objects would cause the connector to fail and shutdown.
# If 'all', then any errors during data transformations or reading objects would have no impact and the connector and

> # worker task will continue with next records.

Reviewer: maybe just add errors are logged and ignored for 'all'
